### PR TITLE
[WIP] Make dockerfile simple and reduce the build time and the size

### PR DESCRIPTION
**Pull request by kchygoe**
## Motivation and Context

- Docker image was so big (more than 7.3GB) and this can be reduced easily
- Docker build time was too longer than neccesary
- We have only one Dockerfile and the structure can be simple
- `blueoil` command doesn't work on current image(because not rehased in pyenv)

## Description

- Place Dockerfile on repository root to clean the structure
- Reduce the docker image size to around 5.3GB by followings
  - Change base image to cuda-cudnn-runtime from cuda-cudnn-devel
  - Stop to use pyenv and use OS default python3
  - Change base OS version from 16.04 to 18.04 to use python3.6.x
- Make it easier to switch python version(use `--build-arg UBUNTU_VERSION=16.04`)
- Make `blueoil` command works fine

## How has this been tested?

Build locally and manually run commands

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature / Optimization (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by kchygoe** [comment]
dlk: GLIBCのバージョンまわり確認・修正が必要

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+# Build coco. It needs numpy.
+COPY lmnet/third_party /home/blueoil/lmnet/third_party
+# https://github.com/cocodataset/cocoapi/blob/440d145a30b410a2a6032827c568cff5dc1d2abf/PythonAPI/setup.py#L2
+RUN cd /home/blueoil/lmnet/third_party/coco/PythonAPI && pip install -e .
+
+# For development
+RUN apt-get install -y --no-install-recommends \
+    x11-apps \
+    imagemagick \
+    ttf-dejavu \
```
Comment (hadusam) [inline Dockerfile]: why do we need `ttf-dejavu`?

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+# Build coco. It needs numpy.
+COPY lmnet/third_party /home/blueoil/lmnet/third_party
+# https://github.com/cocodataset/cocoapi/blob/440d145a30b410a2a6032827c568cff5dc1d2abf/PythonAPI/setup.py#L2
+RUN cd /home/blueoil/lmnet/third_party/coco/PythonAPI && pip install -e .
+
+# For development
+RUN apt-get install -y --no-install-recommends \
+    x11-apps \
+    imagemagick \
```
Comment (hadusam) [inline Dockerfile]: We may not need `x11-apps` and `imagemagick` (they are for development), we can remove them from runtime docker image.

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+FROM nvidia/cuda:10.0-cudnn7-runtime-ubuntu${UBUNTU_VERSION}
+# python3.6 is provided by ubuntu18.04
+# python3.5 is provided by ubuntu16.04
+
+MAINTAINER blueoil-admin@leapmind.io
+
+# TensorBoard
+EXPOSE 6006
+
+ENV LD_LIBRARY_PATH /usr/local/cuda/extras/CUPTI/lib64:${LD_LIBRARY_PATH}
```
Comment (hadusam) [inline Dockerfile]: I think we may be able to remove this ENV.

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+RUN locale-gen en_US.UTF-8
+ENV LANG en_US.UTF-8
+ENV LANGUAGE en_US:en
+ENV LC_ALL en_US.UTF-8
+
+RUN ln -s /usr/bin/python3 /usr/bin/python
+RUN ln -s /usr/bin/pip3 /usr/bin/pip
+
+# Install dependencies for Pillow, Scipy and matplotlib for display
+RUN apt-get install -y --no-install-recommends \
```
Comment (hadusam) [inline Dockerfile]: I think it is better to merge this to [line:15](https://github.com/blue-oil/blueoil/pull/374/files#diff-3254677a7917c6c01f55212f86c57fbfR15).
Please change to run `apt-get update` and `apt-get clean && rm -rf /var/lib/apt/lists/*` only once to reduce build time and layer size.

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    libjpeg8-dev \
+    liblapack-dev \
+    llvm \
+    make \
+    python3-matplotlib \
+    python3-pil \
+    wget
+
+# install aarch64 cross compile environment
+RUN apt-get install -y --no-install-recommends crossbuild-essential-arm64
```
Comment (hadusam) [inline Dockerfile]: I think it is better to merge this to [line:15](https://github.com/blue-oil/blueoil/pull/374/files#diff-3254677a7917c6c01f55212f86c57fbfR15).

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+# Install x-compiler
+RUN apt-get install -y g++-5-arm-linux-gnueabihf && \
```
Comment (hadusam) [inline Dockerfile]: I think it is better to merge this to [line:15](https://github.com/blue-oil/blueoil/pull/374/files#diff-3254677a7917c6c01f55212f86c57fbfR15).
And separate `ln -s` part.

**Comment by iizukak** [comment]
@kchygoe 
I build docker image with `make build` @ `18e46feed92a`.
And i god

```
REPOSITORY                                            TAG                               IMAGE ID            CREATED             SIZE
blueoil_iizuka                                        v0.8.0-21-g377227f                18e46feed92a        2 minutes ago       6.34GB
```

Image size is 6.34GB. Your comment says 5.3GB. There is 1.04GB difference. It's correct ?

File: Dockerfile [hunk 7]
```diff
@@ -0,0 +1,94 @@
+ARG UBUNTU_VERSION=18.04
+FROM nvidia/cuda:10.0-cudnn7-runtime-ubuntu${UBUNTU_VERSION}
+# python3.6 is provided by ubuntu18.04
+# python3.5 is provided by ubuntu16.04
+
+MAINTAINER blueoil-admin@leapmind.io
```
Comment (hadusam) [inline Dockerfile]: Change to use `LABEL`
```suggestion
LABEL maintainer "LeapMind Inc. <blueoil-admin@leapmind.io>"
```

File: Dockerfile (hunk 1 above)
Comment (kchygoe) [reply Dockerfile]: `lmnet/lmnet/visualize.py:31:FONT = "DejaVuSans.ttf"`

File: Dockerfile (hunk 4 above)
Comment (kchygoe) [reply Dockerfile]: ?? This PR is to reduce `apt update` and `apt clean`. But I guess we can merge apt lines as you mention except `locale` settings

**Comment by iizukak** [comment]
@kchygoe 
What's current status of this PR?
Now Fixing?

**Comment by kchygoe** [comment]
Now working on dlk issue with the image from ubuntu18.04. It could be another PR. let me change this PR as WIP

**Comment by iizukak** [comment]
@kchygoe Thanks 👍

**Comment by kchygoe** [comment]
I'll work on new base-image in new PR.

**Pull request closed without merging**